src/plugins/discover/public/components/data_types/logs/summary_column/utils.tsx

```diff
@@ -0,0 +1,126 @@
+  const containerName = getFieldValue(doc, constants.CONTAINER_NAME_FIELD);
+  const containerId = getFieldValue(doc, constants.CONTAINER_ID_FIELD);
+  const cloudInstanceId = getFieldValue(doc, constants.CLOUD_INSTANCE_ID_FIELD);
+
+  return {
+    [constants.SERVICE_NAME_FIELD]: serviceName,
+    [constants.HOST_NAME_FIELD]: hostName,
+    [constants.AGENT_NAME_FIELD]: agentName,
+    [constants.ORCHESTRATOR_CLUSTER_NAME_FIELD]: orchestratorClusterName,
+    [constants.ORCHESTRATOR_RESOURCE_ID_FIELD]: orchestratorResourceId,
+    [constants.ORCHESTRATOR_NAMESPACE_FIELD]: orchestratorNamespace,
+    [constants.CONTAINER_NAME_FIELD]: containerName,
+    [constants.CONTAINER_ID_FIELD]: containerId,
+    [constants.CLOUD_INSTANCE_ID_FIELD]: cloudInstanceId,
+  };
+};
+
+/**
+ * createResourceFields definitions
+ */
+const AgentIcon = dynamic(() => import('@kbn/custom-icons/src/components/agent_icon'));
+
+const resourceCustomComponentsMap: Partial<
+  Record<keyof ResourceFields, React.ComponentType<FieldBadgeWithActionsProps>>
+> = {
+  [constants.SERVICE_NAME_FIELD]: ServiceNameBadgeWithActions,
+};
+
+export interface ResourceFieldDescriptor {
+  ResourceBadge: React.ComponentType<FieldBadgeWithActionsProps>;
+  Icon?: () => JSX.Element;
+  name: keyof ResourceFields;
+  value: string;
+}
+
+export const createResourceFields = (row: DataTableRecord): ResourceFieldDescriptor[] => {
+  const resourceDoc = getUnformattedResourceFields(row as LogDocument);
+
+  const availableResourceFields = getAvailableResourceFields(resourceDoc);
+
+  const resourceFields = availableResourceFields.map((name) => ({
+    name,
+    value: resourceDoc[name] as string,
+    ResourceBadge: resourceCustomComponentsMap[name] ?? FieldBadgeWithActions,
+    ...(name === constants.SERVICE_NAME_FIELD && {
+      Icon: () => (
+        <AgentIcon
+          agentName={resourceDoc[constants.AGENT_NAME_FIELD] as AgentName}
+          size="m"
+          css={css`
+            margin-right: ${euiThemeVars.euiSizeXS};
+          `}
+        />
+      ),
+    }),
+  }));
+
+  return resourceFields;
+};
+
+/**
+ * formatJsonDocumentForContent definitions
+ */
+export const formatJsonDocumentForContent = (row: DataTableRecord) => {
+  const flattenedResult: DataTableRecord['flattened'] = {};
+  const rawFieldResult: DataTableRecord['raw']['fields'] = {};
+  const { raw, flattened } = row;
+  const { fields } = raw;
+
+  // We need 2 loops here for flattened and raw.fields. Flattened contains all fields,
+  // whereas raw.fields only contains certain fields excluding _ignored
+  for (const fieldName in flattened) {
+    if (isFieldAllowed(fieldName) && flattened[fieldName]) {
+      flattenedResult[fieldName] = flattened[fieldName];
+    }
+  }
+
+  for (const fieldName in fields) {
+    if (isFieldAllowed(fieldName) && fields[fieldName]) {
+      rawFieldResult[fieldName] = fields[fieldName];
+    }
+  }
```

Comment:
I agree 100% with your concern here, it might become a heavy task to process when it has many fields.
This is how it used to work with virtual columns too, so I guess we would have the same performance issue there.
As an initial improvement, I refactored its usage in the Content sub-component to only format the document when we need it, so in case none of the defined fields for the content exists and we have to fallback to the whole document.

This should at least mitigate the number of documents we have to process this way and, in an ideal scenario, reduce the need for it to zero.

[perf(discover): optimize document formatting to remove properties](https://github.com/elastic/kibana/pull/192567/commits/933b08013c589ab0c6ba4e9f61257daff24446ac)

P.S. I also searched for any possible improvement at the low level code, bug given we have to assert each property against a prefix and not a the full name, it seems there is not much margin for improvement.